### PR TITLE
Implements preliminary version of Moodle's backup/restore feature

### PR DESCRIPTION
**Pull request by juho-jaakkola**
This is already able to make a backup of the database and restore it afterwards.
- [ ] Figure out why image files and other media do not get restored from the backup package

### CODE REVIEW

**Comment by juho-jaakkola** [comment]
I've rebased this on the latest master, and updated the issue description.

File: backup/moodle2/backup_hvp_activity_task.class.php [hunk 1]
```diff
@@ -40,9 +40,31 @@ protected function define_my_settings() {
...
+        // TODO Should the paths like this:
+        //     {"path":"images/imageSlideBackground-56c5a34e5743e.jpg","mime"
+        // be converted here into format:
+        //     {"path":"@@PLUGINFILE@@/image.jpg","mime"
+        // or something similar so that the backup system's annotate_files()
+        // method would be able to include them into the backup?
+
+        // This regex already matches all jpg files in such way:
+        $search = "/\"path\":\"images\/([^\"]+)\",\"mime\"/";
+        $content = preg_replace($search, '@@PLUGINFILE@@/${1}","mime"', $content);
```
Comment (juho-jaakkola) [inline backup/moodle2/backup_hvp_activity_task.class.php]: Here I tried replacing this:
```
"path":"images/imageSlideBackground-56c5a34e5743e.jpg"
```
with this:
```
"path":"@@PLUGINFILE@@/imageSlideBackground-56c5a34e5743e.jpg"
```
but it has no effect on the backup. There's either something I'm missing, or I just don't really understand how the backup feature is supposed to include the files.

**Comment by juho-jaakkola** [comment]
I've also asked help from the Moodle community: https://moodle.org/mod/forum/discuss.php?d=334643

File: backup/moodle2/backup_hvp_activity_task.class.php (hunk 1 above)
Comment (juho-jaakkola) [inline backup/moodle2/backup_hvp_activity_task.class.php]: I've most likely misunderstood this part. Apparently the backup feature does not check existence of files in the text fields, but instead fetches them from the `mdl_files` table directly.

**Comment by juho-jaakkola** [comment]
I have today:
 - Rebased this on latest master
 - Created a follow issue for finishing the work: #107
 - Marked the backup feature as unsupported (`FEATURE_BACKUP_MOODLE2: false`) so no-one accidentally uses the feature and ends up with a broken backup

So I think it is fine to merge this, and get back to finishing the missing parts later.

File: backup/moodle2/backup_hvp_stepslib.php [hunk 2]
```diff
@@ -15,68 +15,107 @@
 // along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
 /**
- * @package moodlecore
- * @subpackage backup-moodle2
- * @copyright 2010 onwards Eloy Lafuente (stronk7) {@link http://stronk7.com}
+ * @package mod_hvp
+ * @copyright 2016 Mediamaisteri Oy {@link http://www.mediamaisteri.com}
```
Comment (juho-jaakkola) [inline backup/moodle2/backup_hvp_stepslib.php]: I'm not an OS lawyer, so I'm not completely sure who should be mentioned on this line. :P Perhaps there should be at least a mention of the original author? Or perhaps this isn't really a problem in the first place.

File: backup/moodle2/backup_hvp_stepslib.php [hunk 3]
```diff
@@ -15,68 +15,107 @@
...
+            'license',
+            'meta_keywords',
+            'meta_description',
+            'filtered',
+            'slug',
+            'timecreated',
+            'timemodified',
+        ));
+
+            $hvp_libraries = new backup_nested_element('libraries');
```
Comment (fnoks) [inline backup/moodle2/backup_hvp_stepslib.php]: Please fix indentation

File: backup/moodle2/backup_hvp_stepslib.php [hunk 4]
```diff
@@ -15,68 +15,107 @@
...
+                    'machine_name',
+                    'major_version',
+                    'minor_version',
+                    'patch_version',
+                    'dependency_type',
+                    'drop_css',
+                    'weight',
+                ));
+
+            $content_user_data_entries = new backup_nested_element('content_user_data');
```
Comment (fnoks) [inline backup/moodle2/backup_hvp_stepslib.php]: Please fix indentation

File: backup/moodle2/backup_hvp_stepslib.php [hunk 5]
```diff
@@ -15,68 +15,107 @@
...
         // Build the tree
-        $choice->add_child($options);
-        $options->add_child($option);
+        $hvp->add_child($content_user_data_entries);
+            $content_user_data_entries->add_child($content_user_data);
 
-        $choice->add_child($answers);
-        $answers->add_child($answer);
+        $hvp->add_child($hvp_libraries);
+            $hvp_libraries->add_child($hvp_library);
```
Comment (fnoks) [inline backup/moodle2/backup_hvp_stepslib.php]: Please fix indentation

File: backup/moodle2/backup_hvp_stepslib.php [hunk 6]
```diff
@@ -15,68 +15,107 @@
...
+                    'data',
+                    'preloaded',
+                    'delete_on_content_change',
+                ));
 
         // Build the tree
-        $choice->add_child($options);
-        $options->add_child($option);
+        $hvp->add_child($content_user_data_entries);
+            $content_user_data_entries->add_child($content_user_data);
```
Comment (fnoks) [inline backup/moodle2/backup_hvp_stepslib.php]: Please fix indentation

File: backup/moodle2/backup_hvp_stepslib.php [hunk 7]
```diff
@@ -15,68 +15,107 @@
...
+        $hvp->add_child($hvp_libraries);
+            $hvp_libraries->add_child($hvp_library);
 
         // Define sources
-        $choice->set_source_table('choice', array('id' => backup::VAR_ACTIVITYID));
-
-        $option->set_source_table('choice_options', array('choiceid' => backup::VAR_PARENTID), 'id ASC');
+        $hvp->set_source_table('hvp', array('id' => backup::VAR_ACTIVITYID));
+
+            $lib_sql = "SELECT
```
Comment (fnoks) [inline backup/moodle2/backup_hvp_stepslib.php]: Please fix indentation

File: backup/moodle2/backup_hvp_stepslib.php [hunk 8]
```diff
@@ -15,68 +15,107 @@
...
+                    drop_css,
+                    weight
+                FROM {hvp_libraries} hl
+                JOIN {hvp_contents_libraries} hcl
+                    ON hl.id = hcl.library_id
+                JOIN {hvp} h
+                    ON h.id = hcl.hvp_id
+                WHERE h.id = ?";
+
+            $hvp_library->set_source_sql($lib_sql, array(backup::VAR_ACTIVITYID));
```
Comment (fnoks) [inline backup/moodle2/backup_hvp_stepslib.php]: Please fix indentation

File: backup/moodle2/backup_hvp_stepslib.php (hunk 7 above)
Comment (juho-jaakkola) [inline backup/moodle2/backup_hvp_stepslib.php]: The indentations are deliberate. They mimic the way how the elements are nested within the backup structure.

But it's true that this isn't part of Moodle's coding conventions. I can align them if you want.

**Comment by juho-jaakkola** [comment]
The previously indented lines have now been aligned.

**Comment by vaughany** [comment]
:+1:

**Comment by vaughany** [comment]
Hi everyone.  I'm keen to get a handle on where the H5P Moodle plugin is regarding backup and restore, mainly because I'm going to need it very soon.

Having a look at @juho-jaakkola 's PR it seems that all of it already exists within the H5P Moodle plugin master branch, despite this PR not being merged. I can see that the backup works but attached media doesn't (I had a H5P consisting solely of a pie chart which worked fine, but a collage in which the data loaded but the images obviously failed). 

As my employer loves H5P and we're going to be using it a lot rather soon, I offer my services (and that of another developer @cgreenwood - we're _that keen_ to get this working!) to get backup and restore working, if this still a community effort.

**Comment by icc** [comment]
Hm, not sure why the PR isn't merged. All the changes should be available on master.

Any effort to try and fix the restore media/file issue is very welcome. I had set off some time to debug the exported files, but other issues had to come first.

**Comment by vaughany** [comment]
Any pointers or info you could give us to get us going in the right direction? I'm not really sure what the issues even are yet. I'm coming at this from almost no knowledge of Moodle's backup and restore code, only the knowledge that we really need it working! :)

**Comment by icc** [comment]
I'm sorry, my knowledge on the issue is quite limited as well. My guess is that there's something wrong when creating the export.

**Comment by vaughany** [comment]
Thanks, @icc. 

Do you have any pointers @juho-jaakkola ?

**Comment by vaughany** [comment]
Please note that @cgreenwood and I have implemented a solution and submitted a PR: #131.

**Comment by icc** [comment]
Great! I'll have a look and do some testing.

**Pull request closed without merging**